File: D star Lite C++/D_start_Lite/src/Initialize.cpp
/*
 * Academic License - for use in teaching, academic research, and meeting
 * course requirements at degree granting institutions only.  Not for
 * government, commercial, or other organizational use.
 *
 * Initialize.cpp
 *
 * Code generation for function 'Initialize'
 *
 */

/* Include files */
#include "main.h"
#include "Initialize.h"
#include "GetBoundary.h"
#include "GetObstacleCover.h"


/* Parameters Declarations */
const double INF = 999.0;
const int robot_size = 45; //45cm as width
const int accuracy = 10; //10cm as one grid
const int xmax = 81 * 10 / accuracy + 1;
const int ymax = 51 * 10 / accuracy + 1;
double k_m = 0;

/* Matrix Declarations */
Matrix<double, Dynamic, Dynamic> g;
Matrix<double, Dynamic, Dynamic> rhs;
Matrix<double, Dynamic, Dynamic> c;
Matrix<double, Dynamic, Dynamic> U;
Matrix<int, 8, 2> neighbour;

/* Function Definitions */
void Initialize()
{
    g.setConstant(xmax + 1, ymax + 1, INF); // operation cost g
    rhs.setConstant(xmax + 1, ymax + 1, INF); // total cost rhs
    c.setZero(xmax + 1, ymax + 1); // movement cost c, used to determine obstacles
    neighbour << -1, 0, 
                 -1, 1, 
                  0, 1, 
                  1, 1, 
                  1, 0, 
                  1,-1, 
                  0,-1, 
                 -1,-1;
    //GetBoundary();
    GetObstacleCover();
    //标记障碍物
    int row_num = cover.rows();
    for (int i = 0; i < row_num; i++)
    { 
        c(cover(i, 0), cover(i, 1)) = INF;
    }
    //设置目标点rhs为0
    rhs(s_goal(0), s_goal(1)) = 0;
    //初始化U
    U.setZero(1, 4);
    U << s_goal(0), s_goal(1), (s_start - s_goal).norm(), 0;
}

//移除矩阵中指定行
void removeRow(MatrixXd& matrix, unsigned int rowToRemove)
{
    unsigned int numRows = matrix.rows() - 1;
    unsigned int numCols = matrix.cols();

    if (rowToRemove < numRows)
    {
        matrix.block(rowToRemove, 0, numRows - rowToRemove, numCols) = matrix.bottomRows(numRows - rowToRemove);
    }

    matrix.conservativeResize(numRows, numCols);
}

//移除矩阵中指定列
void removeColumn(MatrixXd& matrix, unsigned int colToRemove)
{
    unsigned int numRows = matrix.rows();
    unsigned int numCols = matrix.cols() - 1;

    if (colToRemove < numCols)
    {
        matrix.block(0, colToRemove, numRows, numCols - colToRemove) = matrix.rightCols(numCols - colToRemove);
    }

    matrix.conservativeResize(numRows, numCols);
}

//在矩阵最后一行插入向量
void appendRow(MatrixXd& matrix, RowVectorXd& vector)
{
    unsigned int numRows = matrix.rows() + 1;

    matrix.conservativeResize(numRows, NoChange);

    matrix.row(numRows-1) = vector;
}

/* End of code generation (Initialize.cpp) */

File: D_star Lite/D_star Lite/Initialize.m
function Initialize()
global accuracy;
% input
global xmax;
global ymax;
global U; % open list
global k_m;
global g;
global c;
global rhs;
global neighbour;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% INITIALIZATION %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
xmax=81*10/accuracy+1;%定义地图大小-行
ymax=51*10/accuracy+1;%定义地图大小-列
%%
U=[];%优先级列表，用于存储待扩展的非一致节点(g(s)!=rhs(s))
k_m=0;%记录最初起点到当前起始点的代价值
g=[];%g和rhs表示节点s到目标点的最小代价的估计值
rhs=[];%rhs是由其前向节点（起始点到当前点）的g值计算得到
neighbour=[-1,0; % 8 direction
           -1,1;
           0,1;
           1,1;
           1,0;
           1,-1;
           0,-1;
           -1,-1]; 
g(1:xmax,1:ymax)=Inf;%遍历地图节点集S并初始化,这里注意行列对应的坐标是相反的
rhs(1:xmax,1:ymax)=Inf;
c(1:xmax,1:ymax)=0;
end